File: scripts/MemoryGame.js
import State from "./State.js";
import End from "./End.js";
import shuffleArray from "./Util.js";

export default class MemoryGame {
    constructor() {
        this.rows = 0;
        this.cols = 0;
        this.board = null;
        this.numOfPlays = 0;
        this.firstCell = null;    
    }
    getBoard() {
        let clone = JSON.parse(JSON.stringify(this.board));
        for (let i = 0; i < this.rows; i++) {
            for (let j = 0; j < this.cols; j++) {
                switch (clone[i][j].show) {
                    case State.HIDDEN:
                        clone[i][j] = State.HIDDEN;
                        break;
                    case State.SHOW:
                        clone[i][j] = clone[i][j].value;
                        break;
                    case State.INVISIBLE:
                        clone[i][j] = State.INVISIBLE;
                        break;
                }
            }
        }
        return clone;
    }
    init(n) {
        this.numOfPlays = 0;
        let { rows, cols } = this.computeMatrixDimensions(n * 2);
        this.rows = rows;
        this.cols = cols;
        this.board = Array(rows).fill().map(() => Array(cols).fill());
        let array = Array.from(Array(n).keys());
        let set = array.concat(array);
        shuffleArray(set);
        for (let i = 0; i < rows; i++) {
            for (let j = 0; j < cols; j++) {
                this.board[i][j] = { value: set[cols * i + j], show: State.HIDDEN };
            }
        }
        return this.getBoard();
    }
    play({ x, y }) {
        let cell = this.board[x][y];
        if (cell.show === State.HIDDEN) {
            cell.show = State.SHOW;
            if (!this.firstCell) {
                this.firstCell = cell;
                return { end: End.NO, plays: this.numOfPlays, card1: cell, card2: null, show: false };
            } else {
                this.numOfPlays++;
                if (this.firstCell.value === cell.value) {
                    cell.show = State.INVISIBLE;
                    this.firstCell.show = State.INVISIBLE;
                } else {
                    cell.show = State.HIDDEN;
                    this.firstCell.show = State.HIDDEN;
                }
                let c1 = this.firstCell, c2 = cell;
                this.firstCell = null;
                return { end: this.endOfGame(), plays: this.numOfPlays, card1: c1, card2: c2, show: c1.show === State.INVISIBLE };
            }
        }
        return { end: End.NO, plays: this.numOfPlays, card1: null, card2: null, show: false };
    }
    endOfGame() {
        return this.board.flat().every(v => v.show === State.INVISIBLE) ? End.YES : End.NO;
    }
    computeMatrixDimensions(size) {
        let row = Math.floor(Math.sqrt(size));
        let col = Math.floor(size / row);
        while (row * col !== size) {
            col = Math.floor(size / ++row);
        }
        return { rows: Math.min(row, col), cols: Math.max(row, col) };
    }
}

File: src/main/java/model/MemoryGame.java
package model;

import java.util.Arrays;
import java.util.Random;
import java.util.stream.IntStream;

public class MemoryGame {
    private int rows;
    private int cols;
    private Data[][] board;
    private Data firstCell;
    private Player turn;
    private int score1;
    private int score2;

    public MemoryGame() {
        this.rows = 0;
        this.cols = 0;
        this.board = null;
        this.firstCell = null;
        this.turn = Player.PLAYER1;
        this.score1 = 0;
        this.score2 = 0;
    }

    public Player getTurn() {
        return turn;
    }

    public int[] getScores() {
        return new int[] { this.score1, this.score2 };
    }

    public Data[][] getBoard() {
        var clone = new Data[rows][cols];
        for (var i = 0; i < this.rows; i++) {
            for (var j = 0; j < this.cols; j++) {
                switch (this.board[i][j].getShow()) {
                    case State.HIDDEN:
                        clone[i][j] = new Data(-1, State.HIDDEN, new Cell(i, j));
                        break;
                    case State.SHOW:
                        clone[i][j] = new Data(board[i][j].getValue(), State.SHOW, new Cell(i, j));
                        break;
                    case State.INVISIBLE:
                        clone[i][j] = new Data(-1, State.INVISIBLE, new Cell(i, j));
                        break;
                }
            }
        }
        return clone;
    }

    private Cell computeMatrixDimensions(int size) {
        var row = Math.floor(Math.sqrt(size));
        var col = Math.floor(size / row);
        while (row * col != size) {
            col = Math.floor(size / ++row);
        }
        return new Cell((int) Math.min(row, col), (int) Math.max(row, col));
    }

    public Data[][] init(int n) {
        var cell = this.computeMatrixDimensions(n * 2);
        this.rows = cell.x();
        this.cols = cell.y();
        this.board = new Data[this.rows][this.cols];
        var array = IntStream.rangeClosed(1, n).toArray();
        var set = IntStream.concat(Arrays.stream(array), Arrays.stream(array)).toArray();
        shuffleArray(set);
        for (var i = 0; i < rows; i++) {
            for (var j = 0; j < cols; j++) {
                this.board[i][j] = new Data(set[cols * i + j], State.HIDDEN, new Cell(i, j));
            }
        }
        return this.getBoard();
    }

    private void shuffleArray(int[] array) {
        Random rand = new Random();
        for (int i = 0; i < array.length; i++) {
            int randomIndexToSwap = rand.nextInt(array.length);
            int temp = array[randomIndexToSwap];
            array[randomIndexToSwap] = array[i];
            array[i] = temp;
        }
    }

    public Winner getWinner() {
        if (this.board != null && Arrays.stream(this.board).flatMap(o -> Arrays.stream(o))
                .allMatch(v -> v.getShow() == State.INVISIBLE)) {
            return this.score1 > this.score2 ? Winner.PLAYER1
                    : this.score1 < this.score2 ? Winner.PLAYER2 : Winner.DRAW;
        }
        return Winner.NONE;
    }

    public Result play(Player player, Cell coords) {
        if (player != this.turn) {
            throw new IllegalArgumentException("It is not your turn.");
        }
        Data cell = this.board[coords.x()][coords.y()];
        if (cell.getShow() == State.INVISIBLE) {
            throw new IllegalArgumentException("This card was already discovered.");
        }
        if (cell.getShow() == State.SHOW) {
            throw new IllegalArgumentException("This card is already open.");
        }
        cell.setShow(State.SHOW);
        if (this.firstCell == null) {
            this.firstCell = cell;
            return new Result(cell, null, false);
        } else {
            if (this.firstCell.getValue() == cell.getValue()) {
                if (this.turn == Player.PLAYER1) {
                    this.score1++;
                } else {
                    this.score2++;
                }
                cell.setShow(State.INVISIBLE);
                this.firstCell.setShow(State.INVISIBLE);
            } else {
                cell.setShow(State.HIDDEN);
                this.firstCell.setShow(State.HIDDEN);
                this.turn = this.turn == Player.PLAYER1 ? Player.PLAYER2 : Player.PLAYER1;
            }
            var c1 = this.firstCell;
            var c2 = cell;
            this.firstCell = null;
            return new Result(c1, c2, cell.getShow() == State.INVISIBLE);
        }
    }
}
